src/fides/api/ctl/routes/router_factory.py

```diff
@@ -0,0 +1,392 @@
+    API_PREFIX,
+    CLI_SCOPE_PREFIX_MAPPING,
+    forbid_if_default,
+    forbid_if_editing_any_is_default,
+    forbid_if_editing_is_default,
+)
+from fides.api.ctl.sql_models import (
+    DataCategory,
+    models_with_default_field,
+    sql_model_map,
+)
+from fides.api.ctl.utils import errors
+from fides.api.ctl.utils.api_router import APIRouter
+from fides.api.ops.api.v1.scope_registry import CREATE, DELETE, READ, UPDATE
+from fides.api.ops.oauth.utils import verify_oauth_client_prod
+from fides.api.ops.schemas.dataset import validate_data_categories_against_db
+
+
+async def get_data_categories_from_db(async_session: AsyncSession) -> List[FidesKey]:
+    """Similar method to one on the ops side except this uses an async session to retrieve data categories"""
+    resources = await list_resource(DataCategory, async_session)
+    data_categories = [res.fides_key for res in resources]
+    return data_categories
+
+
+async def validate_data_categories(
+    resource: FidesModel, async_session: AsyncSession
+) -> None:
+    """Validate data categories defined on Datasets against data categories in the db"""
+    if not isinstance(resource, Dataset):
+        return
+
+    try:
+        defined_data_categories: List[FidesKey] = await get_data_categories_from_db(
+            async_session
+        )
+        validate_data_categories_against_db(resource, defined_data_categories)
+    except PydanticValidationError as e:
+        raise HTTPException(
+            status_code=HTTP_422_UNPROCESSABLE_ENTITY, detail=e.errors()
+        )
+
+
+def generic_router_factory(fides_model: FidesModelType, model_type: str) -> APIRouter:
+    """
+    Compose all of the individual route factories into a single coherent Router.
+    """
+
+    object_router = APIRouter()
+
+    create_router = create_router_factory(
+        fides_model=fides_model, model_type=model_type
+    )
+    list_router = list_router_factory(fides_model=fides_model, model_type=model_type)
+    get_router = get_router_factory(fides_model=fides_model, model_type=model_type)
+    delete_router = delete_router_factory(
+        fides_model=fides_model, model_type=model_type
+    )
+    update_router = update_router_factory(
+        fides_model=fides_model, model_type=model_type
+    )
+    upsert_router = upsert_router_factory(
+        fides_model=fides_model, model_type=model_type
+    )
+
+    object_router.include_router(create_router)
+    object_router.include_router(list_router)
+    object_router.include_router(get_router)
+    object_router.include_router(delete_router)
+    object_router.include_router(update_router)
+    object_router.include_router(upsert_router)
+
+    return object_router
+
+
+def create_router_factory(fides_model: FidesModelType, model_type: str) -> APIRouter:
+    """Return a configured version of a generic 'Create' route."""
+
+    router = APIRouter(prefix=f"{API_PREFIX}/{model_type}", tags=[fides_model.__name__])
```

Comment:
when I tried to create API routes and then add them, it didn't respect the prefix, so I ended up having to create and return APIRouter objects instead of APIRoutes

This method is slightly more verbose but works all the same